# Brazillian E-commerce Investigation
Import neccessary libraries

# Objectives
Analyse historical Brazilian Olist Online Store Data .
Identify trends, patterns, and key insights from the dataset
Develop predictive models to forecast future rankings
Prepare a cleaned dataset for visualisations in Tableau ....

# Inputs
'olist_customers_dataset.csv' 
'olist_geolocation_dataset.csv'
'olist_order_items_dataset.csv'
'olist_order_payments_dataset.csv'
'olist_order_reviews_dataset.csv'
'olist_products_dataset.csv'
'olist_sellers_dataset.csv'
'product_category_name_translation.csv'

# Outputs
1.Initial merged dataset: 'order_items_merged.csv'


# Additional Comments
Ensure data integrity by handling missing values and duplicates
Ethical considerations include recognising data biases and cultural influences in ranking trends

To analyse the Brazilian Olist online store  dataset effectively, the following steps are taken:

1. Data Understanding and Preparation
Each CSV files dataset is checked for null values and duplicates, ensuring data integrity
Attributes are standardised to maintain consistency
The 'zip address' column are converted to a five digits to facilitate time series analysis

### Set a new directory

In [24]:
import os
current_dir = os.getcwd()
current_dir

'/Users/anita/Documents/vscode-projects/capstone-analysis/jupyter_notebooks'

### Import Libraries

In [25]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px


### Read all the CSV files

In [26]:
csv_files = {
'customers': 'olist_customers_dataset.csv',
'orders': 'olist_orders_dataset.csv',
'order_items': 'olist_order_items_dataset.csv',
'order_payments': 'olist_order_payments_dataset.csv',
'order_reviews': 'olist_order_reviews_dataset.csv',
'products': 'olist_products_dataset.csv',
'sellers': 'olist_sellers_dataset.csv',
'geolocation': 'olist_geolocation_dataset.csv',
'category_translation': 'product_category_name_translation.csv'
}


dataframes = {}
for name, filename in csv_files.items():
    dataframes[name] = pd.read_csv(f"../data/brazilian_csv/{filename}")
    print(f"Loaded {name}: {dataframes[name].shape}")


df_customers = dataframes['customers']
df_orders = dataframes['orders']
df_order_items = dataframes['order_items']
df_order_payments = dataframes['order_payments']
df_order_reviews = dataframes['order_reviews']
df_products = dataframes['products']
df_sellers = dataframes['sellers']
df_geolocation = dataframes['geolocation']
df_category_translation = dataframes['category_translation']

Loaded customers: (99441, 5)
Loaded orders: (99441, 8)
Loaded order_items: (112650, 7)
Loaded order_payments: (103886, 5)
Loaded order_reviews: (99224, 7)
Loaded products: (32951, 9)
Loaded sellers: (3095, 4)
Loaded geolocation: (1000163, 5)
Loaded category_translation: (71, 2)


### Inspect data

In [27]:
df_customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [28]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


There are four objects in the customer data. There is one integer. There is no null value.

In [29]:
df_geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [30]:
df_geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


There is one integer in geolocation data.  There are two floats and two objects. There is no null value.

In [31]:
df_order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [32]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


There are four objects in order item data. There are two floats and one integer. There is no null value.

In [33]:
df_order_items.describe()

,order_item_id,price,freight_value
count,112650.000000,112650.000000,112650.000000
mean,1.197834,120.653739,19.990320
std,0.705124,183.633928,15.806405
min,1.000000,0.850000,0.000000
25%,1.000000,39.900000,13.080000
50%,1.000000,74.990000,16.260000
75%,1.000000,134.900000,21.150000
max,21.000000,6735.000000,409.680000


The mean / average price of the order items is 121 BRL. 
The minimum price is 0.85 BRL
The maximum price is 6735 BRL

The mean / average freight value for the order items is 20 BRL
The minimum fright value is 0 BRL which indicates that there is free option of shipping available.
The maximum freight value is 410 BRL



In [34]:
df_order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [35]:
df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


There are two objects in order payments data. There are one float and two integers. There is no null value.

In [36]:
df_order_payments.describe()

,payment_sequential,payment_installments,payment_value
count,103886.000000,103886.000000,103886.000000
mean,1.092679,2.853349,154.100380
std,0.706584,2.687051,217.494064
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,56.790000
50%,1.000000,1.000000,100.000000
75%,1.000000,4.000000,171.837500
max,29.000000,24.000000,13664.080000



The mean / average price of the order payment sequential is 1. 
The minimum payment sequential is 1
The maximum payment sequential is 29
Most orders are paid in single attempt.


The mean / average payment instalments for the order items is 3
The minimum payment values is 0  which indicates that there is free or failed paymemnts or data issues
The maximum payment values is 13554 BRL

The minimum value of 0 for both payment installments and payment values may include is free products, refunds or missing data


In [37]:
df_order_reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [38]:
df_order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


There are 6 objects and 1 integer. Column 3 and 4 contain null values.

In [39]:
df_order_reviews.describe()

,review_score
count,99224.000000
mean,4.086421
std,1.347579
min,1.000000
25%,4.000000
50%,5.000000
75%,5.000000
max,5.000000


The mean score of order reviews is 4. 
The minimum score of order reviews is 1.
The maximum score of order reviews is 5.

In [40]:
df_orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [41]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


Column 4, 5 and 6 contain null values and all coloumns are object data type.

In [42]:
df_products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [43]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


All columns except column 1 which is product_id contain null values.
There are 7 floats and 2 objects.

In [44]:
df_sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [45]:
df_sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


There is no null values in sellers data. There are 3 objects and 1 integer.

In [46]:
df_category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [47]:
df_category_translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


There is no null value in category translation. They are all objects.

---

### Data Cleaning 

In [48]:
# Check for duplicates in each dataset
# List of all your dataframes with their names
datasets = {
    "df_customers": df_customers,
    "df_geolocation": df_geolocation,
    "df_order_items": df_order_items,
    "df_order_payments": df_order_payments,
    "df_order_reviews": df_order_reviews,
    "df_orders": df_orders,
    "df_products": df_products,
    "df_sellers": df_sellers,
    "df_category_translation": df_category_translation
}


for name, df in datasets.items():
    print(f"{name} - Shape: {df.shape}, Duplicates: {df.duplicated().sum()}")


df_customers - Shape: (99441, 5), Duplicates: 0
df_geolocation - Shape: (1000163, 5), Duplicates: 261831
df_order_items - Shape: (112650, 7), Duplicates: 0
df_order_payments - Shape: (103886, 5), Duplicates: 0
df_order_reviews - Shape: (99224, 7), Duplicates: 0
df_orders - Shape: (99441, 8), Duplicates: 0
df_products - Shape: (32951, 9), Duplicates: 0
df_sellers - Shape: (3095, 4), Duplicates: 0
df_category_translation - Shape: (71, 2), Duplicates: 0


In [49]:
# Check missing values
for name, df in datasets.items():
    print(f"\n{name} Missing Values:")
    print(df.isnull().sum())


df_customers Missing Values:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

df_geolocation Missing Values:
geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

df_order_items Missing Values:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

df_order_payments Missing Values:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

df_order_reviews Missing Values:
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247


In [50]:
# Don't drop these - missing comments are normal customer behavior!
# Just fill with "No Comment" if needed for analysis
df_order_reviews['review_comment_title'].fillna('No Title', inplace=True)
df_order_reviews['review_comment_message'].fillna('No Message', inplace=True)

In [51]:
# Drop rows with any missing values
df_orders = df_orders.dropna()


In [52]:
# 1. Category name - use 'Unknown' or most frequent category
df_products['product_category_name'].fillna('outros', inplace=True) # 'others' in Portuguese

# 2. Numeric features - use median (robust to outliers)
numeric_cols = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty',
'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

for col in numeric_cols:
    median_value = df_products[col].median()
    df_products[col].fillna(median_value, inplace=True)
    print(f"Filled {col} with median: {median_value}")

Filled product_name_lenght with median: 51.0
Filled product_description_lenght with median: 595.0
Filled product_photos_qty with median: 1.0
Filled product_weight_g with median: 700.0
Filled product_length_cm with median: 25.0
Filled product_height_cm with median: 13.0
Filled product_width_cm with median: 20.0


--- 

### Data Preprocessing

In [53]:
datetime_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

# Convert each column to datetime
for col in datetime_columns:
    df_orders[col] = pd.to_datetime(df_orders[col], errors='coerce')  # errors='coerce' will set invalid parsing as NaT

# Optional: check data types to confirm
print(df_orders[datetime_columns].dtypes)

order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object


/var/folders/hf/46nj1q753jx0d3gy7yg3ns6c0000gn/T/ipykernel_11257/1992088276.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_orders[col] = pd.to_datetime(df_orders[col], errors='coerce')  # errors='coerce' will set invalid parsing as NaT
/var/folders/hf/46nj1q753jx0d3gy7yg3ns6c0000gn/T/ipykernel_11257/1992088276.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_orders[col] = pd.to_datetime(df_orders[col], errors='coerce')  # errors='coerce' will set invalid parsing as NaT
/var/folders/hf/46nj1q

In [54]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
Index: 96461 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       96461 non-null  object        
 1   customer_id                    96461 non-null  object        
 2   order_status                   96461 non-null  object        
 3   order_purchase_timestamp       96461 non-null  datetime64[ns]
 4   order_approved_at              96461 non-null  datetime64[ns]
 5   order_delivered_carrier_date   96461 non-null  datetime64[ns]
 6   order_delivered_customer_date  96461 non-null  datetime64[ns]
 7   order_estimated_delivery_date  96461 non-null  datetime64[ns]
dtypes: datetime64[ns](5), object(3)
memory usage: 6.6+ MB


In [55]:
print("Creating Master Dataset...")

# Merge orders with customers
master_df = df_orders.merge(df_customers, on='customer_id', how='left')

# Merge with order items
master_df = master_df.merge(df_order_items, on='order_id', how='left')

# Merge with products
master_df = master_df.merge(df_products, on='product_id', how='left')

# Merge with English product category names
master_df = master_df.merge(df_category_translation,
                           on='product_category_name', how='left')

# Merge with order payments
order_payments_agg = df_order_payments.groupby('order_id').agg({
    'payment_sequential': 'max',
    'payment_type': lambda x: ', '.join(x.unique()),
    'payment_installments': 'mean',
    'payment_value': 'sum'
}).reset_index()

master_df = master_df.merge(order_payments_agg, on='order_id', how='left')

# Merge with seller and geolocation data
master_df = master_df.merge(df_sellers, on='seller_id', how='left', suffixes=('', '_seller'))

# Add customer geolocation
customer_geo = df_geolocation.groupby('geolocation_zip_code_prefix').agg({
    'geolocation_lat': 'mean',
    'geolocation_lng': 'mean',
    'geolocation_city': 'first',
    'geolocation_state': 'first'
}).reset_index()

master_df = master_df.merge(customer_geo,
                           left_on='customer_zip_code_prefix',
                           right_on='geolocation_zip_code_prefix',
                           how='left', suffixes=('', '_customer'))

print(f"Master dataset shape: {master_df.shape}")
print(f"Columns: {master_df.columns.tolist()}")

Creating Master Dataset...
Master dataset shape: (110180, 39)
Columns: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm', 'product_category_name_english', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value', 'seller_zip_code_prefix', 'seller_city', 'seller_state', 'geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']


In [56]:
master_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,payment_installments,payment_value,seller_zip_code_prefix,seller_city,seller_state,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,1.0,38.71,9350,maua,SP,3149.0,-23.576983,-46.587161,sao paulo,SP
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,1.0,141.46,31570,belo horizonte,SP,47813.0,-12.177924,-44.660711,barreiras,BA
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,3.0,179.12,14840,guariba,SP,75265.0,-16.745150,-48.514783,vianopolis,GO
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,...,1.0,72.20,31842,belo horizonte,MG,59296.0,-5.774190,-35.271143,sao goncalo do amarante,RN
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,1.0,28.62,8752,mogi das cruzes,SP,9195.0,-23.676370,-46.514627,santo andre,SP


In [57]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110180 entries, 0 to 110179
Data columns (total 39 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       110180 non-null  object        
 1   customer_id                    110180 non-null  object        
 2   order_status                   110180 non-null  object        
 3   order_purchase_timestamp       110180 non-null  datetime64[ns]
 4   order_approved_at              110180 non-null  datetime64[ns]
 5   order_delivered_carrier_date   110180 non-null  datetime64[ns]
 6   order_delivered_customer_date  110180 non-null  datetime64[ns]
 7   order_estimated_delivery_date  110180 non-null  datetime64[ns]
 8   customer_unique_id             110180 non-null  object        
 9   customer_zip_code_prefix       110180 non-null  int64         
 10  customer_city                  110180 non-null  object        
 11  

In [58]:
datetime_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date',
    'shipping_limit_date',

]

# Convert each column to datetime
for col in datetime_columns:
    master_df[col] = pd.to_datetime(master_df[col], errors='coerce')  # errors='coerce' will set invalid parsing as NaT

# Optional: check data types to confirm
print(master_df[datetime_columns].dtypes)

order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
shipping_limit_date              datetime64[ns]
dtype: object


In [59]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110180 entries, 0 to 110179
Data columns (total 39 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       110180 non-null  object        
 1   customer_id                    110180 non-null  object        
 2   order_status                   110180 non-null  object        
 3   order_purchase_timestamp       110180 non-null  datetime64[ns]
 4   order_approved_at              110180 non-null  datetime64[ns]
 5   order_delivered_carrier_date   110180 non-null  datetime64[ns]
 6   order_delivered_customer_date  110180 non-null  datetime64[ns]
 7   order_estimated_delivery_date  110180 non-null  datetime64[ns]
 8   customer_unique_id             110180 non-null  object        
 9   customer_zip_code_prefix       110180 non-null  int64         
 10  customer_city                  110180 non-null  object        
 11  

---

### Feature Engineering

In [60]:
print("\nFeature Engineering...")

# Date features
master_df['order_year'] = master_df['order_purchase_timestamp'].dt.year
master_df['order_month'] = master_df['order_purchase_timestamp'].dt.month
master_df['order_dayofweek'] = master_df['order_purchase_timestamp'].dt.dayofweek
master_df['order_hour'] = master_df['order_purchase_timestamp'].dt.hour

# Calculate delivery time
master_df['delivery_days'] = (master_df['order_delivered_customer_date'] -
                             master_df['order_purchase_timestamp']).dt.days

# Product metrics
master_df['product_volume'] = (master_df['product_length_cm'] *
                              master_df['product_height_cm'] *
                              master_df['product_width_cm'])

# Price per gram
master_df['price_per_gram'] = master_df['price'] / (master_df['product_weight_g'] + 1)

# Order-level aggregations for RFM analysis
print("Creating customer-level aggregations...")

# Get the reference date (latest date in dataset)
reference_date = master_df['order_purchase_timestamp'].max()

customer_metrics = master_df.groupby('customer_unique_id').agg({
    'order_purchase_timestamp': ['count', 'max'],
    'payment_value': ['sum', 'mean'],
    'delivery_days': 'mean',
    'price': ['sum', 'mean'],
    'product_category_name_english': lambda x: x.nunique(),
    'seller_id': lambda x: x.nunique()
}).round(2)

# Flatten column names
customer_metrics.columns = ['_'.join(col).strip() for col in customer_metrics.columns]
customer_metrics = customer_metrics.reset_index()

# Calculate Recency, Frequency, Monetary for RFM
customer_metrics['recency'] = (reference_date - customer_metrics['order_purchase_timestamp_max']).dt.days
customer_metrics['frequency'] = customer_metrics['order_purchase_timestamp_count']
customer_metrics['monetary'] = customer_metrics['payment_value_sum']


Feature Engineering...
Creating customer-level aggregations...


---

### Exploratory Data Analysis

Visuals such as line pl